ckanext/storepublisher/controllers/ui_controller.py

```diff
@@ -56,42 +79,119 @@ def publish(self, id, offering_info=None, errors=None):
+                result.append({x: i[x] for x in requiredFields})
+            for elem in result:
+                elem['text'] = elem.pop('name')
+                elem['value'] = elem.pop('id')
+            return result
+
+        def _validateVersion(version):
+            ver = version
+            if not ver:
+                ver = '1.0'
+            if re.search(r'\.$', ver) is not None:
+                ver += '0'
+            if re.search(r'\.{2,}', ver) is not None:
+                ver = re.sub(r'\.{2,}', r'\.', ver)
+            if re.search(r'^\.', ver) is not None:
+                ver = "1" + ver
+            return ver
+
+        if request.GET:
+            self._listOfTags, self._catRelatives = _sort_categories(self._get_Content('category'))
+            self._listOfCatalogs = self._get_Content('catalog')
+
+            c.offering = {
+                'categories': _getList(self._listOfTags),
+                'catalogs': _getList(self._listOfCatalogs)
+            }
         # when the data is provided
         if request.POST:
             offering_info = {}
             offering_info['pkg_id'] = request.POST.get('pkg_id', '')
             offering_info['name'] = request.POST.get('name', '')
             offering_info['description'] = request.POST.get('description', '')
+            offering_info['license_title'] = request.POST.get(
+                'license_title', '')
+            offering_info['license_description'] = request.POST.get(
+                'license_description', '')
+            offering_info['version'] = _validateVersion(
+                request.POST.get('version', ''))
+            offering_info['is_open'] = 'open' in request.POST
             offering_info['license_title'] = request.POST.get('license_title', '')
             offering_info['license_description'] = request.POST.get('license_description', '')
-            offering_info['version'] = request.POST.get('version', '')
-            offering_info['is_open'] = 'open' in request.POST
+            categories = request.POST.getall('categories')
+            tempList = []
+
+            # Insert all parents in the set until there are no more new parents
+            for cat in categories:
+                tempList.append(self._catRelatives[cat])
+                tempCat = self._catRelatives[cat]
+                while 'parentId' in tempCat and tempCat['parentId']:
+                    tempList.append(self._catRelatives[tempCat['parentId']])
+                    tempCat = self._catRelatives[tempCat['parentId']]
+
+            for cat in tempList:
+                if 'parentId' in cat:
+                    del cat['parentId']
 
-            # Get tags
-            # ''.split(',') ==> ['']
-            tag_string = request.POST.get('tag_string', '')
-            offering_info['tags'] = [] if tag_string == '' else tag_string.split(',')
+            tempList = [dict(tupleized) for tupleized in set(tuple(item.items()) for item in tempList)]
```

Comment:
When adding all ancestors of the selected categories, each of two brothers categories will had once their parent and so on.